### PR TITLE
xds: parse Listener update as xDS v3 resource

### PR DESCRIPTION
**Pull request by dapengzhang0**
This is part of xDS v3 support as per go/grpc-xds-v3-support

In this PR:
- still only send v2 requests to xDS server (No v3 bootstrap or env flag support)
- parse Listener update as v3 proto
- Refactor SDS's Listener watcher to use enovy v3 API
- still parse other resources as v2 proto.

### CODE REVIEW

**Comment by dapengzhang0** [comment]
cc @voidzcy

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java [hunk 1]
```diff
@@ -553,6 +555,9 @@ private void handleLdsResponseForConfigUpdate(DiscoveryResponse ldsResponse) {
     List<String> listenerNames = new ArrayList<>(ldsResponse.getResourcesCount());
     try {
       for (com.google.protobuf.Any res : ldsResponse.getResourcesList()) {
+        if (!res.is(Listener.class)) {
```
Comment (voidzcy) [inline xds/src/main/java/io/grpc/xds/XdsClientImpl.java:558]: Can you use fully qualified name for types that you want to distinguish in v2 and v3? It's hard to follow as there are some inconsistencies (e.g., constant `ADS_TYPE_URL_LDS` without v3 suffix refers to v2 type while `Listener` refers to v3 type).

Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:558]: What will be better is to use v3 unqualified but v2 as qualified (since it will be phased out at some point) wherever possible. So in this case `Listener` is correct but `ADS_TYPE_URL_LDS` should be v3 unqualified.

Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:558]: v3 imported classes should be unqualified and v2 ones can be qualified (since they will go away earlier). Constants like `ADS_TYPE_URL_LDS` are defined here: the v3 ones should have no suffix and the existing v2 ones can be renamed with `_V2` suffix for the same reason.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java [hunk 2]
```diff
@@ -87,6 +87,8 @@
 
   @VisibleForTesting
   static final String ADS_TYPE_URL_LDS = "type.googleapis.com/envoy.api.v2.Listener";
+  private static final String ADS_TYPE_URL_LDS_V3 =
```
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/XdsClientImpl.java:90]: If `ADS_TYPE_URL_LDS` needs to be visible for testing how come `ADS_TYPE_URL_LDS_V3` doesn't need to be if the test is also migrated to v3?

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java [hunk 3]
```diff
@@ -730,7 +738,7 @@ private boolean isRequestedListener(Listener listener) {
         && hasMatchingFilter(listener.getFilterChainsList());
   }
 
-  private boolean isAddressMatching(Address address) {
+  private boolean isAddressMatching(io.envoyproxy.envoy.config.core.v3.Address address) {
```
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/XdsClientImpl.java:741]: Can we change the import so we don't need the v3 fully qualified name?

File: xds/src/main/java/io/grpc/xds/internal/sds/FileBasedPluginCredential.java [hunk 4]
```diff
@@ -77,6 +78,19 @@
     secretData = buildDataSourceFromConfigStruct(value.getStructValue());
   }
 
+  FileBasedPluginCredential(
```
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/internal/sds/FileBasedPluginCredential.java:81]: Just wondering - what is the need to keep both the v2 and v3 versions? When everything is moved to v3 are we going to remove the v2 version of the constructor?

File: xds/src/main/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManager.java [hunk 5]
```diff
@@ -223,15 +225,16 @@ void verifySubjectAltNameInChain(X509Certificate[] peerCertChain) throws Certifi
...
-    List<String> verifyList = certContext.getVerifySubjectAltNameList();
+    List<StringMatcher> verifyList = certContext.getMatchSubjectAltNamesList();
     if (verifyList.isEmpty()) {
       return;
     }
     if (peerCertChain == null || peerCertChain.length < 1) {
       throw new CertificateException("Peer certificate(s) missing");
     }
     // verify SANs only in the top cert (leaf cert)
-    verifySubjectAltNameInLeaf(peerCertChain[0], verifyList);
```
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManager.java:234]: Can you comment this instead of deleting the call? Since the comment above is still there.

File: xds/src/main/java/io/grpc/xds/internal/sds/FileBasedPluginCredential.java (hunk 4 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/internal/sds/FileBasedPluginCredential.java:81]: I just kept the existing code unchanged, otherwise the whole class will be almost empty after removing v2 code and with v3 code unimplemented. We don't need v2 version for this class after migration.

File: xds/src/test/java/io/grpc/xds/EnvoyServerProtoDataTest.java [hunk 6]
```diff
@@ -73,8 +73,7 @@ public void listener_convertFromListenerProto() throws InvalidProtocolBufferExce
     assertThat(outFilterChainMatch.getApplicationProtocols()).isEmpty();
     assertThat(outFilterChainMatch.getPrefixRanges()).isEmpty();
     assertThat(outFilter.getDownstreamTlsContext())
-        .isEqualTo(DownstreamTlsContext.fromEnvoyProtoDownstreamTlsContext(
-                io.envoyproxy.envoy.api.v2.auth.DownstreamTlsContext.getDefaultInstance()));
+        .isNull();
```
Comment (sanjaypujare) [inline xds/src/test/java/io/grpc/xds/EnvoyServerProtoDataTest.java:76]: Good catch :-)

File: xds/src/test/java/io/grpc/xds/EnvoyServerProtoDataTest.java [hunk 7]
```diff
@@ -147,7 +119,8 @@ private static FilterChain createInFilter() {
                     .addApplicationProtocols("managed-mtls")
                     .build())
             .setTransportSocket(TransportSocket.newBuilder().setName("tls")
-                .setTypedConfig(Any.pack(CommonTlsContextTestsUtil.buildTestDownstreamTlsContext()))
+                .setTypedConfig(
+                    Any.pack(CommonTlsContextTestsUtil.buildTestDownstreamTlsContextV2()))
```
Comment (sanjaypujare) [inline xds/src/test/java/io/grpc/xds/EnvoyServerProtoDataTest.java:123]: Why `V2` ? With all Listener related stuff migrating to v3 why can't this be v3 as well?

File: xds/src/test/java/io/grpc/xds/XdsClientImplTest.java [hunk 8]
```diff
@@ -1435,6 +1435,7 @@ public void cdsResponseWithUpstreamTlsContext() {
     StreamObserver<DiscoveryRequest> requestObserver = requestObservers.poll();
 
     // Management server sends back CDS response with UpstreamTlsContext.
+    // Server still using envoy API V2.
```
Comment (sanjaypujare) [inline xds/src/test/java/io/grpc/xds/XdsClientImplTest.java:1438]: Not sure I understand this. Does it need a TODO comment so this can be cleaned up later?

File: xds/src/test/java/io/grpc/xds/EnvoyServerProtoDataTest.java (hunk 7 above)
Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/EnvoyServerProtoDataTest.java:123]: Good catch. I missed that.

File: xds/src/test/java/io/grpc/xds/XdsClientImplTest.java (hunk 8 above)
Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/XdsClientImplTest.java:1438]: Currently the server only support v2, and we will support v2 servers for at least a year. I think we don't need clean up in the near future.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 1 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: Checking type name for `ADS_TYPE_URL_LDS_V2` instead, which might be easier to understand.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 2 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: Done.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 3 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: Done.

File: xds/src/main/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManager.java (hunk 5 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManager.java:234]: Done.

File: xds/src/test/java/io/grpc/xds/XdsClientImplTest.java (hunk 8 above)
Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/XdsClientImplTest.java:1438]: I will remove this comment here.

File: xds/src/main/java/io/grpc/xds/internal/sds/FileBasedPluginCredential.java (hunk 4 above)
Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/internal/sds/FileBasedPluginCredential.java:81]: ok

File: xds/src/test/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManagerTest.java [hunk 9]
```diff
@@ -84,180 +83,6 @@ public void emptySanListContextTest() throws CertificateException, IOException {
     trustManager.verifySubjectAltNameInChain(certs);
   }
 
-  @Test
```
Comment (sanjaypujare) [inline xds/src/test/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManagerTest.java:87]: Looking at the latest findings in https://github.com/grpc/grpc-java/issues/7166#issuecomment-652536045 where we will use match_subject_alt_names with EXACT_MATCH most of these tests will be applicable (after making the change). Would it be better to `@Ignore` them instead of just deleting them?

File: xds/src/test/java/io/grpc/xds/internal/sds/CommonTlsContextTestsUtil.java [hunk 10]
```diff
@@ -48,18 +53,83 @@
...
+          io.envoyproxy.envoy.api.v2.auth.SdsSecretConfig.newBuilder()
               .setName(name)
-              .setSdsConfig(SdsClientTest.buildConfigSource(targetUri, channelType))
+              .setSdsConfig(buildConfigSourceV2(targetUri, channelType))
               .build();
     }
     return sdsSecretConfig;
   }
 
+  private static io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.SdsSecretConfig
```
Comment (sanjaypujare) [inline xds/src/test/java/io/grpc/xds/internal/sds/CommonTlsContextTestsUtil.java:69]: With the import this type doesn't need to be fully qualified, correct?

File: xds/src/test/java/io/grpc/xds/XdsClientImplTest.java [hunk 11]
```diff
@@ -336,7 +336,7 @@ public void ldsResponseWithoutMatchingResource() {
     // Client sends an LDS request for the host name (with port) to management server.
     verify(requestObserver)
         .onNext(eq(buildDiscoveryRequest(NODE, "", TARGET_AUTHORITY,
-            XdsClientImpl.ADS_TYPE_URL_LDS, "")));
+            XdsClientImpl.ADS_TYPE_URL_LDS_v2, "")));
```
Comment (sanjaypujare) [inline xds/src/test/java/io/grpc/xds/XdsClientImplTest.java:339]: A high level comment about tests. So by renaming `ADS_TYPE_URL_LDS` to `ADS_TYPE_URL_LDS_v2` we are only testing/validating v2 responses. What about testing v3 responses? Is there a separate PR for it or a TODO comment to indicate it needs to be done?

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java [hunk 12]
```diff
@@ -86,7 +86,9 @@
   static final int INITIAL_RESOURCE_FETCH_TIMEOUT_SEC = 15;
 
   @VisibleForTesting
-  static final String ADS_TYPE_URL_LDS = "type.googleapis.com/envoy.api.v2.Listener";
+  static final String ADS_TYPE_URL_LDS_v2 = "type.googleapis.com/envoy.api.v2.Listener";
```
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/XdsClientImpl.java:89]: Is the lowercase `v2` intended or typo? A const literal should be all uppercase, right? Same thing for other `_v2` literals below.

File: xds/src/test/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManagerTest.java (hunk 9 above)
Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManagerTest.java:87]: I used `@Ignore` for some other tests. These tests here can not compile because `addVerifySubjectAltName()` does not exist for v3 api.

File: xds/src/test/java/io/grpc/xds/XdsClientImplTest.java (hunk 11 above)
Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/XdsClientImplTest.java:339]: Currently we are still sending v2 request, so responses are expected to be v2. Testing v3 response will be after v3 bootstrap and env flag enabled.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java [hunk 13]
```diff
@@ -553,14 +555,17 @@ private void handleLdsResponseForConfigUpdate(DiscoveryResponse ldsResponse) {
     List<String> listenerNames = new ArrayList<>(ldsResponse.getResourcesCount());
     try {
       for (com.google.protobuf.Any res : ldsResponse.getResourcesList()) {
+        if (res.getTypeUrl().equals(ADS_TYPE_URL_LDS_v2)) {
```
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/XdsClientImpl.java:558]: Couple of things:
- based on https://github.com/envoyproxy/envoy/blob/master/api/envoy/service/discovery/v3/discovery.proto#L97 which says "Type URL ... Must be consistent with the type_url in the 'resources' repeated Any". Should we have that check here now that v2 and v3 are the 2 possibilities?
- shouldn't the typeUrl in ACK/NACK match what was there in the response? I think it is possible that the client asks for v3 in the request but the response comes back with v2. In that case should ACK/NACK match the typeUrl of the original request or the response?

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java [hunk 14]
```diff
@@ -553,14 +555,17 @@ private void handleLdsResponseForConfigUpdate(DiscoveryResponse ldsResponse) {
...
+        }
         Listener listener = res.unpack(Listener.class);
         listeners.add(listener);
         listenerNames.add(listener.getName());
       }
     } catch (InvalidProtocolBufferException e) {
       logger.log(XdsLogLevel.WARNING, "Failed to unpack Listeners in LDS response {0}", e);
       adsStream.sendNackRequest(
-          ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName),
+          ADS_TYPE_URL_LDS_v2, ImmutableList.of(ldsResourceName),
```
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/XdsClientImpl.java:568]: My comment above about v2 vs v3. For now this is okay but when the v3 migration is complete and our request uses v3 typeUrl should we analyze the responses for v2 vs v3 typeUrl and use that for resource validation and also subsequent ACK/NACK?

File: xds/src/test/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManagerTest.java (hunk 9 above)
Comment (sanjaypujare) [reply xds/src/test/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManagerTest.java:87]: Sounds good. Can you add a TODO comment (assigned to me) to remind that these 11(?) tests need to be brought back when the new `match_subject_alt_names` field is implemented?

File: xds/src/test/java/io/grpc/xds/XdsClientImplTest.java (hunk 11 above)
Comment (sanjaypujare) [reply xds/src/test/java/io/grpc/xds/XdsClientImplTest.java:339]: Yes, I realized it later. Sounds good.

File: xds/src/test/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManagerTest.java (hunk 9 above)
Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManagerTest.java:87]: I added a TOOD referring the issue here https://github.com/grpc/grpc-java/pull/7169/files/6f32fb0eec0589cd4d606418de35ca2a85cfb855#diff-9efb11763826a58994b5e41a318c58d8R54 which has discussions on `match_subject_alt_names`

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 14 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:568]: My understanding of the design doc is we should accept (ACK) both v2 and v3 response before we drop v2 support.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 13 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:558]: I did have the check that v2 and v3 are the 2 possibilities (Will ACK for 2 possibilities)
https://github.com/grpc/grpc-java/pull/7169/files/6f32fb0eec0589cd4d606418de35ca2a85cfb855#diff-2ab746208e54edd7bcbba5fc408b6946R1327

But I do not check the type_url of response is consistent with that of repeated Any - I assume the TD is consistent.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 14 above)
Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:568]: Agreed but my point was slightly different. When the v3 migration is complete we will request v3 resources by using the v3 type Url in the DiscoveryRequest (https://github.com/envoyproxy/envoy/blob/master/api/envoy/service/discovery/v3/discovery.proto#L54). Let's say the control plane decides to return v2 resources instead (which is possible and that's why we need to maintain v2 support for some time). So in the DiscoveryResponse the typeUrl (https://github.com/envoyproxy/envoy/blob/master/api/envoy/service/discovery/v3/discovery.proto#L98) will be for v2 and so will be in the individual resources (https://github.com/envoyproxy/envoy/blob/master/api/envoy/service/discovery/v3/discovery.proto#L78). When that happens and you decide to ACK or NACK that response will that ACK/NACK use v2 typeUrl (to match the response) or the v3 typeUrl (to match the original request)? If the answer it to match the original request then there isn't much to do but if the answer is to match the response then we need additional logic (may be not in this PR).

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:568]: So in the case of (after migration, before dropping v2 support) xds client sending v3 request and receiving v2 response (and v2 individual resources of course): we parse the v2 resources to v3 resource objects and ACK the v2 response (with v2 [typeUrl](https://github.com/envoyproxy/envoy/blob/master/api/envoy/service/discovery/v3/discovery.proto#L98)  in the ACK I think. cc @markdroth to confirm) if no error occurs.

File: xds/src/test/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManagerTest.java (hunk 9 above)
Comment (sanjaypujare) [reply xds/src/test/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManagerTest.java:87]: Okay I added an item in the issue so I can track it there.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 14 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:568]: Basically I think we should ACK/NACK with type_url version the same as the type_url of the response.

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:568]: Basically I think we should ACK/NACK with type_url version the same as the type_url of the response.

Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: In that case this line can be changed to 
```
ldsResponse.getTypeUrl(), ImmutableList.of(ldsResourceName),
```
and similar changes elsewhere for all ACK/NACK cases. Provided we confirm that the typeUrl in ACK/NACK has to match the one in the response. WDYT?

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 13 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: >In that case should ACK/NACK match the typeUrl of the original request or the response?

I think we should match the response.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 14 above)
Comment (markdroth) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: Good question.  I think we will want the ACK/NACK to use the version that we originally requested, not the version that the server sent back.  This is because the xDS protocol does not have a completely separate ACK/NACK message; the ACK or NACK is implicitly included in the next request, which may also change the set of requested resource names at the same time as it sends an ACK/NACK.  And since we are already expecting the server to be understanding that the v2 and v3 resource names are the same, that logic should not be any different for ACK/NACK than it should be for any other request.

@htuch, can you give me a sanity check that the above is correct?

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 12 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: Corrected.

File: xds/src/test/java/io/grpc/xds/internal/sds/CommonTlsContextTestsUtil.java (hunk 10 above)
Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/internal/sds/CommonTlsContextTestsUtil.java]: Right. Done.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 14 above)
Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: @markdroth I can see it either way. If the server uses the typeUrl (from the response) for its bookkeeping and version tracking of resources for clients then ACKing/NACKing with the original request's typeUrl could confuse it. It will be good to get a definitive answer from the xDS spec. BTW, what does C-core do?

Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: In any case, using the typeUrl of the original request sounds more logical so let's go ahead with that for now. If we hear otherwise we can make that change as a separate PR.

Comment (markdroth) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: My work-in-progress PR for C-core (grpc/grpc#23281) will always use the version that the client is using, not the version that the server sends back.

The more I think about this, the less sense it makes to me to use the version that the server sent back.  Any server that decides to send a different version than the one that was asked for is already needing to understand the mapping between those two versions.  And having the client switch which version it's referring to in the middle of the stream seems like it could badly confuse the xDS semantics, because (as I said earlier) an ACK/NACK can be part of the same request as a change to the set of subscribed resource names, and the client will have already previously sent a request for a different set of resource names using the other version.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 13 above)
Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: > > In that case should ACK/NACK match the typeUrl of the original request or the response?
> 
> I think we should match the response.

That discussion is in a separate thread so we can continue it there. I agree that there is no need to validate responses from the control plane (TD).

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java [hunk 15]
```diff
@@ -553,14 +555,17 @@ private void handleLdsResponseForConfigUpdate(DiscoveryResponse ldsResponse) {
     List<String> listenerNames = new ArrayList<>(ldsResponse.getResourcesCount());
     try {
       for (com.google.protobuf.Any res : ldsResponse.getResourcesList()) {
+        if (res.getTypeUrl().equals(ADS_TYPE_URL_LDS_V2)) {
+          res = res.toBuilder().setTypeUrl(ADS_TYPE_URL_LDS).build();
```
Comment (voidzcy) [inline xds/src/main/java/io/grpc/xds/XdsClientImpl.java:559]: Is this trick really gonna work?

Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:559]: I think the answer is yes since that is the whole basis for this PR (or this set of PRs). Having a unit test to prove it works should not be difficult.

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:559]: Why not?

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 14 above)
Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: I agree (with Mark's comment) and go ahead with that approach. Just to add for better/wider understanding, the server side should ignore the typeUrls in the ACK/NACK but keep using the typeUrl used to establish the stream (in my opinion).

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 15 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:559]: I am suspecting `Any.unpack` would break. I really doubt `Any.unpack` can deserialize a v3 Listener bytes to a v2 Listener without screwing up the message. Do you have any doc reference doing that?

Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:559]: @voidzcy the logic here is the opposite. We are trying to deserialize a v2 proto payload as v3. And the approach has been discussed and agreed to for this work (in the design doc as well as a pending gRFC). If the field numbers are maintained and proper handling is done for deprecated and new fields I don't see why it won't work.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:559]: Ok, I see. They put placeholders for deprecated fields and arrange fields in the way that the message is ensured to be the same in the wire format as v2. Then this should work.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 14 above)
Comment (htuch) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: Yeah, ACK/NACK with the original version.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 15 above)
Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:559]: resolving...

File: xds/src/test/java/io/grpc/xds/internal/sds/CommonTlsContextTestsUtil.java [hunk 16]
```diff
@@ -18,14 +18,19 @@
 
 import com.google.common.base.Strings;
 import com.google.protobuf.BoolValue;
-import io.envoyproxy.envoy.api.v2.auth.CertificateValidationContext;
-import io.envoyproxy.envoy.api.v2.auth.CommonTlsContext;
-import io.envoyproxy.envoy.api.v2.auth.CommonTlsContext.CombinedCertificateValidationContext;
```
Comment (sanjaypujare) [inline xds/src/test/java/io/grpc/xds/internal/sds/CommonTlsContextTestsUtil.java:23]: I didn't see the correspnding v3 import for CombinedCertificateValidationContext. It will be good to have that

Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/internal/sds/CommonTlsContextTestsUtil.java:23]: Done.
